## Homework 3- Feature Engineering for Classification

Goal: Train the best classifier possible for Heart Disease Prediction while trying various feature engineering techniques and learning to run experiments to find the best possible model.


### Feature Preprocessing
For each categorical features implement two feature representations:
1. OneHot Encoding: For example, transform `X_train['Sex']` with values `M or F` into two features `X_Train['Sex_M']`, `X_Train['Sex_F']` with values `0 or 1`
2. Target Encoding: For example, transform `X_train['Sex']` with values `M or F` into a feature `X_Train['Sex-TargetEncoded']` with value equal to the average rate of heart disease of `M and F` respectively.

Please implement these yourself, but you can check against sklearn.preprocessing implementations for correctness.

The set of categorical features is:
`categorical_features = ['Sex', 'ChestPainType', 'RestingECG', 'ExerciseAngina', 'ST_Slope']`


For numerical features implement two feature normalisations:
1. Standard Scaler (scale numeric values to have zero mean and unit variance). For example `X_train['Age-Scaled'] = (X_train['Age'] - mean) / standard_deviation`
2. MinMax Normalization (subtract min and divide by max - min). For example `X_train['Age-MinMax'] = (X_train['Age'] - min) / (max - min)`

Please implement these yourself but you can check against the sklearn.preprocessing implementations.

The set of numeric features is:
`numeric_features = ["Age", "RestingBP", "Cholesterol", "FastingBS", "MaxHR", "Oldpeak"]`
    
Note, any feature preprocessing parameters (like mean or standard deviation) should be calculated on training data only.

### Feature Engineering:
Create at least 5 custom features as functions of other features:

Some ideas:
  - A binary feature representing High Heart Rate `Custom-BinaryHighMaxHR`
  - A bucketized categorical feature of Oldpeak or of MaxHR called `Oldpeak`
  - A feature cross of a bucketized version of OldPeak and MaxHR for example `HighOldPeak_X_LowMaxHR`
  
### Feature Experiments:
Please run at least this set of experiments, but try others as you see fit.
1. All features with no scaling on numeric features and OneHot for categorical. No custom features.
2. All features using StandardScaler for numeric and OneHot for categorical. No custom features.
3. All features using StandardScaler for numeric and TargetEncoding for categorical. No custom features.
4. All features using MinMax-Normalization for numeric and OneHot for categorical. No custom features.
5. The kitchen sink: Include everything to try to get the best performance possible.
6. Only custom features. How good can you get with your own custom feature set?
7. Only categorical features using one of the encodings.
8. Only numerical features using one of the encodings.


### Model Training
Model training Data:
Separate data into a training set with 80% of the samples and a test set with 20%. 
`X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=42)`. Note that ideally we would also have a validation set, but since this dataset is fairly small, we will just do train/test splits. Please use the same random state as above so we have comparable performance across the class.

### Models
Please run all experiments using at least 3 different models
1. Logistic regression using `sklearn.linear_model.LogisticRegression`. Extra credit of +2 if you use your own gradient descent implementation from Homework 2.
2. Decision Trees: using `sklearn.tree.DecisionTreeClassifier`. Hyperparameter here is the depth of the tree.
3. Neural Net classifier using `sklearn.neural_network.MLPClassifier`. Hyperparameters here include the shape of the network and the learning rate. Consider using the 'adam' solver as constructor argument, 'relu' activation fuctions (which is the default), and try with a shape of network, but start small with a network of size `(4,1)`, that is 4 hidden units in one layer or `(3,2)` which includes 2 hidden layers of 3 units each.
4. At least 1 other classifier. Some ideas include SVM classifiers or Gradient Boosted Decision trees which are all available in sklearn as well


### Question 1: Run Experiments (15 points)

We have many varying parameters. Often in ML we run multiple experiments to find the best result.
- at least 8 different feature setups, maybe more
- at least 4 different model algorithms
- some number of hyperparemeters, let's say we try at least 10 for each model

This gives us a minimum of 320 different experiments. To easily keep track of these, let's create a spreadsheet to track the experiment results. 

Create a .csv file with the following columns tracking your experiments. Make sure this .csv can be loaded into Google sheets so I can grade it. Upload the .csv separately.

1. Experiment Name string. For example `Experiment1-LogReg-OneHot-Custom`. This name be anything.
2. List of Feature Included in Model separated by semicolons. Please name these according to this scheme: `<FeatureName>-<Preprocessing>-<Value>` for example `MaxHR-StandardScaler` or `RestingECG-OneHot-Normal`. For custom features name them `Custom-<Name>` for example: `Custom-BucketizedOldpeak`. As an example, a row in this columns might look like `MaxHR; MaxHR-StandardScaler; Custom-BinaryHighHeartRate; ...`. If your features are in your dataframe, you can create this list with `';'.join(X_train.columns)`.
3. Model Type (MyLogisticRegression or DecisionTree or anything else you want to try)
4. Epochs (for logistic regression)
5. Hyperparameters used in your model, for example depth of tree or size of neural network
6. Training Accuracy (w/ threshold of 0.5)
7. Testing Accuracy (w/ threshold of 0.5)
8. Training PR-AUC (Area under the precision/recall curve)
9. Testing PR-AUC (Area under the precision/recall curve)
10. Training Precision (w/ threshold of 0.5)
11. Testing Precision (w/ threshold of 0.5)
12. Training Recall (w/ threshold of 0.5)
13. Testing Recall (w/ threshold of 0.5)

Extra Credit (2 points): Use Weights And Biases (https://wandb.ai/) to track your experiments as an alternative to a spreadsheet (https://towardsdatascience.com/introduction-to-weight-biases-track-and-visualize-your-machine-learning-experiments-in-3-lines-9c9553b0f99d)


### Question 2: Analyze Logistic Regression Experiments (4 points)

- Question 2.1: What is the best set of features and parameters for Logistic Regression in terms of Test PR-AUC. Is this the same as the best in terms of accuracy? Discuss why you think this experiment showed the best results.
- Question 2.2: If features are well normalized, we can get a sense of feature importance by looking at the absolute value of the weights of each feature. Print out the highest 5 weights by absolute value. Which features are these? Discuss what this set of 5 features has on the model.
- Question 2.3: Train a model with just those top 5 features. What is the Test Accuracy and PR-AUC?
- Question 2.4: Describe the custom features you added. Why did you pick these?

### Question 3: Analyze Other Experiments (4 points)

- Question 3.1: What is the best set of features and parameters for Decision Tree classifier in terms of Test PR-AUC.  Discuss why you think this experiment showed the best results.
- Question 3.2: What is the best set of features and parameters for Neural Net classifier in terms of Test PR-AUC.  Discuss why you think this experiment showed the best results.
- Question 3.3: What is the best set of features and parameters for another chosen model in terms of Test PR-AUC.  Discuss why you think this experiment showed the best results.

### Question 4: Achieving particular peformance charactaristics (2 points)
- Question 4.1: From your trained models, can you produce a classifier that has around 90% recall on the test set? How? What is the precision of this model?
- Question 4.2: From your trained models, can you produce a classifier that has around 90% precision on the test set? How? What is the recall of this model?

### Question 5: Taking a step back (2 points)
- Question 5.1: What is your best performing model over all experiments? Describe why you think this was the best? What is the PR-AUC and Accuracy of this model? Feel free to share your results on discord, the best model in the class will receive extra credit of 3 points!
- Question 5.2: Discuss your thoughts on this process? What surprised you? What was hard or tedious? What did you learn?


In [3]:
# Imports
import wandb
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
from matplotlib.pyplot import figure
from sklearn.metrics import accuracy_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, precision_recall_curve, auc

In [4]:
# Load data from our csv
df = pd.read_csv('../../Data/heart.csv')

# Select out Binary and Categorical Features
numeric_features = ["Age", "RestingBP", "Cholesterol", "FastingBS", "MaxHR", "Oldpeak"]
categorical_features = ['Sex', 'ChestPainType', 'RestingECG', 'ExerciseAngina', 'ST_Slope']
label_feature = "HeartDisease"
X = df[numeric_features + categorical_features]
y = df[label_feature]

# Create Train / Test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=42)

# Keeping track of custom features list
custom_features = []

In [5]:
# Examples of Pandas operations you can use:
# Calculate mean of a column:
print(X_train['Age'].mean())
# Calculate standard deviation of a column:
print(X_train['Age'].std())

# Example of creating a custom new feature column, in this case a binary feature indicating age > 50
X_train_old = X_train.copy()
X_test_old = X_test.copy()

53.65122615803815
9.364289800106517


In [6]:
# Feature Preprocessing
# 1. OneHot Encoding: For example, transform X_train['Sex'] with values M or F into two features X_Train['Sex_M'], X_Train['Sex_F'] with values 0 or 1
def oneHotEncoder(features, categories):
    feature_list = []
    for i in categories:
        for j in features[i].unique():
            features[i + "-OneHot-" + j] = (features[i] == j).astype(int)
            feature_list.append(i + "-OneHot-" + j)
    return feature_list

# Outputs all the categorical columns with oneHot Encoding
OneHot_Encoding_Feature_List = oneHotEncoder(X_train, categorical_features) # X_train has been fitted with new columns
oneHotEncoder(X_test, categorical_features) # X_test has been fitted with new columns
X_train[OneHot_Encoding_Feature_List] # Showing the new fitted columns in X_train

Sex-OneHot-M  Sex-OneHot-F  ChestPainType-OneHot-NAP  \
795             1             0                         1   
25              1             0                         1   
84              1             0                         0   
10              0             1                         1   
344             1             0                         0   
..            ...           ...                       ...   
106             0             1                         0   
270             1             0                         0   
860             1             0                         0   
435             1             0                         0   
102             0             1                         0   

     ChestPainType-OneHot-ASY  ChestPainType-OneHot-TA  \
795                         0                        0   
25                          0                        0   
84                          1                        0   
10                          0                        0   
344                         1                        0   
..                        ...                      ...   
106                         1                        0   
270                         1                        0   
860                         1                        0   
435                         1                        0   
102                         1                        0   

     ChestPainType-OneHot-ATA  RestingECG-OneHot-Normal  \
795                         0                         1   
25                          0                         1   
84                          0                         1   
10                          0                         1   
344                         0                         1   
..                        ...                       ...   
106                         0                         0   
270                         0                         1   
860                         0                         1   
435                         0                         0   
102                         0                         1   

     RestingECG-OneHot-LVH  RestingECG-OneHot-ST  ExerciseAngina-OneHot-N  \
795                      0                     0                        1   
25                       0                     0                        1   
84                       0                     0                        0   
10                       0                     0                        1   
344                      0                     0                        1   
..                     ...                   ...                      ...   
106                      0                     1                        1   
270                      0                     0                        1   
860                      0                     0                        0   
435                      0                     1                        0   
102                      0                     0                        1   

     ExerciseAngina-OneHot-Y  ST_Slope-OneHot-Down  ST_Slope-OneHot-Up  \
795                        0                     1                   0   
25                         0                     0                   1   
84                         1                     0                   0   
10                         0                     0                   1   
344                        0                     0                   0   
..                       ...                   ...                 ...   
106                        0                     0                   1   
270                        0                     0                   1   
860                        1                     0                   1   
435                        1                     0                   1   
102                        0                     0                   0   

     ST_Slope-OneHot-Flat  
795                     

In [7]:
# Feature Preprocessing
# 2. Target Encoding: For example, transform X_train['Sex'] with values M or F into a feature X_Train['Sex-TargetEncoded'] with value equal to the average rate of heart disease of M and F respectively.
def targetEncoder(features, label, categories):
    feature_list = []
    for i in categories:
        features[i+"-TargetEncoded"] = features[i].map(pd.concat([features, label], axis=1).groupby(i)[label_feature].mean())
        feature_list.append(i+"-TargetEncoded")
    return feature_list

targetEncoded_feature_List = targetEncoder(X_train, y_train, categorical_features) # X_train has been fitted with new columns
targetEncoder(X_test, y_test, categorical_features)  # X_test has been fitted with new columns
X_train[targetEncoded_feature_List] # Showing the new fitted columns in X_train

Sex-TargetEncoded  ChestPainType-TargetEncoded  RestingECG-TargetEncoded  \
795           0.632042                     0.364198                  0.509009   
25            0.632042                     0.364198                  0.509009   
84            0.632042                     0.770574                  0.509009   
10            0.253012                     0.364198                  0.509009   
344           0.632042                     0.770574                  0.509009   
..                 ...                          ...                       ...   
106           0.253012                     0.770574                  0.645390   
270           0.632042                     0.770574                  0.509009   
860           0.632042                     0.770574                  0.509009   
435           0.632042                     0.770574                  0.645390   
102           0.253012                     0.770574                  0.509009   

     ExerciseAngina-TargetEncoded  ST_Slope-TargetEncoded  
795                      0.327146                0.770833  
25                       0.327146                0.175896  
84                       0.858086                0.817942  
10                       0.327146                0.175896  
344                      0.327146                0.817942  
..                            ...                     ...  
106                      0.327146                0.175896  
270                      0.327146                0.175896  
860                      0.858086                0.175896  
435                      0.858086                0.175896  
102                      0.327146                0.817942  

[734 rows x 5 columns]

In [8]:
# Feature Preprocessing
# 1. Standard Scaler (scale numeric values to have zero mean and unit variance). For example X_train['Age-Scaled'] = (X_train['Age'] - mean) / standard_deviation
def standardScaler(features, categories):
    features_list = []
    for i in categories:
        features[i+"-StandardScaler"] = (features[i] - features[i].mean()) / features[i].std()
        features_list.append(i+"-StandardScaler")
    return features_list

standardScaler_Feature_List = standardScaler(X_train, numeric_features) # X_train has been fitted with new columns
standardScaler(X_test, numeric_features) # X_test has been fitted with new columns
X_train[standardScaler_Feature_List] #  Showing the new fitted columns in X_train

Age-StandardScaler  RestingBP-StandardScaler  Cholesterol-StandardScaler  \
795           -1.244219                 -0.708502                    0.372549   
25            -1.884951                 -0.166172                    0.086087   
84             0.250822                  0.918489                    0.123050   
10            -1.778162                 -0.166172                    0.104569   
344           -0.283121                 -0.708502                   -1.845220   
..                  ...                       ...                         ...   
106           -0.603487                 -0.708502                    0.501919   
270           -0.923853                 -0.708502                    0.233938   
860            0.677977                 -0.166172                    0.492678   
435            0.677977                  1.026955                   -1.845220   
102           -1.457796                  0.918489                    1.777136   

     FastingBS-StandardScaler  MaxHR-StandardScaler  Oldpeak-StandardScaler  
795                  1.841354              2.282796               -0.096995  
25                  -0.542339              1.651116               -0.835717  
84                   1.841354             -0.441327                0.087685  
10                  -0.542339              0.229834               -0.835717  
344                  1.841354             -1.270407               -0.835717  
..                        ...                   ...                     ...  
106                 -0.542339             -1.033527               -0.835717  
270                 -0.542339              0.150874               -0.835717  
860                 -0.542339              0.308794                0.457046  
435                 -0.542339             -0.717687               -0.835717  
102                 -0.542339             -0.243926                1.011087  

[734 rows x 6 columns]

In [9]:
# Feature Preprocessing
# 2. MinMax Normalization (subtract min and divide by max - min). For example X_train['Age-MinMax'] = (X_train['Age'] - min) / (max - min)
def minMax(features, categories):
    features_list = []
    for i in categories:
        features[i+"-MinMax"] = (features[i] - features[i].min()) / (features[i].max() - features[i].min())
        features_list.append(i+"-MinMax")
    return features_list

minMax_feature_list = minMax(X_train, numeric_features) # X_train has been fitted with new columns
minMax(X_test, numeric_features) # X_test has been fitted with new columns
X_train[minMax_feature_list] # Showing the new fitted columns in X_train


Age-MinMax  RestingBP-MinMax  Cholesterol-MinMax  FastingBS-MinMax  \
795    0.270833              0.60            0.398010               1.0   
25     0.145833              0.65            0.346600               0.0   
84     0.562500              0.75            0.353234               1.0   
10     0.166667              0.65            0.349917               0.0   
344    0.458333              0.60            0.000000               1.0   
..          ...               ...                 ...               ...   
106    0.395833              0.60            0.421227               0.0   
270    0.333333              0.60            0.373134               0.0   
860    0.645833              0.65            0.419569               0.0   
435    0.645833              0.76            0.000000               0.0   
102    0.229167              0.75            0.650083               0.0   

     MaxHR-MinMax  Oldpeak-MinMax  
795      0.943662        0.386364  
25       0.830986        0.295455  
84       0.457746        0.409091  
10       0.577465        0.295455  
344      0.309859        0.295455  
..            ...             ...  
106      0.352113        0.295455  
270      0.563380        0.295455  
860      0.591549        0.454545  
435      0.408451        0.295455  
102      0.492958        0.522727  

[734 rows x 6 columns]

In [10]:
# Feature Engineering --> 1st custom feature
# A binary feature representing High Heart Rate BinaryHighMaxHR
def custom_BinaryFeature(features, old_feature, new_feature):
    feature_mean = features[0][old_feature].mean()
    features[0][new_feature] = (features[0][old_feature] >= feature_mean).apply(int)
    features[1][new_feature] = (features[1][old_feature] >= feature_mean).apply(int)
    return new_feature

custom_features.append(custom_BinaryFeature((X_train, X_test), "MaxHR", "Custom-BinaryHighMaxHR")) 
X_train[["MaxHR", "Custom-BinaryHighMaxHR"]].head(10)  # Showing the new fitted columns in X_train

MaxHR  Custom-BinaryHighMaxHR
795    194                       1
25     178                       1
84     125                       0
10     142                       1
344    104                       0
254     96                       0
398    122                       0
244    103                       0
621    142                       1
118    185                       1

In [11]:
# Feature Engineering --> 2nd custom feature 
# A bucketized categorical feature of Oldpeak called bucketized_Oldpeak
def bucketized_feature(features, feature, intervals):
    for f in features:
        f["Custom-BucketizedLow" + feature] = (f[feature] <= intervals[0]).apply(int)
        f["Custom-BucketizedMid" + feature] = ((f[feature] <  intervals[1]) & (f[feature] >  intervals[0])).apply(int)
        f["Custom-BucketizedHigh" + feature] = (f[feature] >= intervals[1]).apply(int)
    return ["Custom-BucketizedLow" + feature, "Custom-BucketizedMid" + feature, "Custom-BucketizedHigh" + feature]


print("Max", X_train["Oldpeak"].max(), "\tMean:", X_train["Oldpeak"].mean(), "\tMin", X_train["Oldpeak"].min(), "\tStd", X_train["Oldpeak"].std())
bucketized_Oldpeak_list = bucketized_feature((X_train, X_test), "Oldpeak", (0, 2))
for i in bucketized_Oldpeak_list: custom_features.append(i)
X_train[bucketized_Oldpeak_list + ["Oldpeak"]].head(10)

Max 6.2 	Mean: 0.9050408719346048 	Min -2.6 	Std 1.0829519487461572


Custom-BucketizedLowOldpeak  Custom-BucketizedMidOldpeak  \
795                            0                            1   
25                             1                            0   
84                             0                            1   
10                             1                            0   
344                            1                            0   
254                            0                            0   
398                            0                            1   
244                            0                            1   
621                            0                            1   
118                            1                            0   

     Custom-BucketizedHighOldpeak  Oldpeak  
795                             0      0.8  
25                              0      0.0  
84                              0      1.0  
10                              0      0.0  
344                             0      0.0  
254                             1      2.0  
398                             0      1.0  
244                             0      1.0  
621                             0      0.6  
118                             0      0.0

In [12]:
# Feature Engineering --> 3rd custom feature
# A bucketized categorical feature of MaxHR called bucketized_MaxHR
print("Max", X_train["MaxHR"].max(), "\tMean:", X_train["MaxHR"].mean(), "\tMin", X_train["MaxHR"].min(), "\tStd", X_train["MaxHR"].std())
bucketized_MaxHR_list = bucketized_feature((X_train, X_test), "MaxHR", (110, 160))
for i in bucketized_MaxHR_list: custom_features.append(i)
X_train[bucketized_MaxHR_list + ["MaxHR"]].head(10)


Max 202 	Mean: 136.17847411444143 	Min 60 	Std 25.329253934908074


Custom-BucketizedLowMaxHR  Custom-BucketizedMidMaxHR  \
795                          0                          0   
25                           0                          0   
84                           0                          1   
10                           0                          1   
344                          1                          0   
254                          1                          0   
398                          0                          1   
244                          1                          0   
621                          0                          1   
118                          0                          0   

     Custom-BucketizedHighMaxHR  MaxHR  
795                           1    194  
25                            1    178  
84                            0    125  
10                            0    142  
344                           0    104  
254                           0     96  
398                           0    122  
244                           0    103  
621                           0    142  
118                           1    185

In [13]:
# Feature Engineering --> 4th custom feature
# A feature cross of a bucketized version of OldPeak and MaxHR for example HighOldpeak_X_LowMaxHR
def cross_feature(features, feature_one, feature_two, new_feature):
    custom_feature = []
    for f in features:
        f[new_feature] = ((f[feature_one] == 1) & (f[feature_two] == 1)).astype(int)
    return new_feature

custom_features.append(cross_feature((X_train, X_test), "Custom-BucketizedHighOldpeak", "Custom-BucketizedLowMaxHR", "Custom-BucketizedHighOldpeak_X_LowMaxHR"))
custom_features.append(cross_feature((X_train, X_test), "Custom-BucketizedMidOldpeak", "Custom-BucketizedMidMaxHR", "Custom-BucketizedMidOldpeak_X_MidMaxHR"))
custom_features.append(cross_feature((X_train, X_test), "Custom-BucketizedLowOldpeak", "Custom-BucketizedHighMaxHR", "Custom-BucketizedLowOldpeak_X_HighMaxHR"))
X_train[["Oldpeak", "MaxHR"] + custom_features[len(custom_features)-4:len(custom_features)-1]].head(10)

Oldpeak  MaxHR  Custom-BucketizedHighMaxHR  \
795      0.8    194                           1   
25       0.0    178                           1   
84       1.0    125                           0   
10       0.0    142                           0   
344      0.0    104                           0   
254      2.0     96                           0   
398      1.0    122                           0   
244      1.0    103                           0   
621      0.6    142                           0   
118      0.0    185                           1   

     Custom-BucketizedHighOldpeak_X_LowMaxHR  \
795                                        0   
25                                         0   
84                                         0   
10                                         0   
344                                        0   
254                                        1   
398                                        0   
244                                        0   
621                                        0   
118                                        0   

     Custom-BucketizedMidOldpeak_X_MidMaxHR  
795                                       0  
25                                        0  
84                                        1  
10                                        0  
344                                       0  
254                                       0  
398                                       1  
244                                       0  
621                                       1  
118                                       0

In [14]:
# Feature Engineering --> 5th custom feature
# A feature cross of a bucketized version of OldPeak and MaxHR for example HighOldpeak_X_LowMaxHR
print("Max", X_train["Age"].max(), "\tMean:", X_train["Age"].mean(), "\tMin", X_train["Age"].min(), "\tStd", X_train["Age"].std())
bucketized_Age_list = bucketized_feature((X_train, X_test), "Age", (45, 60))
for i in bucketized_Age_list: custom_features.append(i)
X_train[["Age"] + bucketized_Age_list].head(10)

Max 77 	Mean: 53.65122615803815 	Min 29 	Std 9.364289800106517


Age  Custom-BucketizedLowAge  Custom-BucketizedMidAge  \
795   42                        1                        0   
25    36                        1                        0   
84    56                        0                        1   
10    37                        1                        0   
344   51                        0                        1   
254   55                        0                        1   
398   52                        0                        1   
244   48                        0                        1   
621   56                        0                        1   
118   35                        1                        0   

     Custom-BucketizedHighAge  
795                         0  
25                          0  
84                          0  
10                          0  
344                         0  
254                         0  
398                         0  
244                         0  
621                         0  
118                         0

In [15]:
# Feature Experiments: All features with no scaling on numeric features and OneHot for categorical. No custom features.
# Experiment 1 - All features with no scaling on numeric features and OneHot for categorical. No custom features.
exp1_train = X_train[numeric_features + OneHot_Encoding_Feature_List]
exp1_test = X_test[numeric_features + OneHot_Encoding_Feature_List]

# Experiment 2 - All features using StandardScaler for numeric and OneHot for categorical. No custom features.
exp2_train = X_train[OneHot_Encoding_Feature_List + standardScaler_Feature_List]
exp2_test = X_test[OneHot_Encoding_Feature_List + standardScaler_Feature_List]

# Experiment 3 - All features using StandardScaler for numeric and TargetEncoding for categorical. No custom features.
exp3_train = X_train[standardScaler_Feature_List +  targetEncoded_feature_List]
exp3_test = X_test[standardScaler_Feature_List + targetEncoded_feature_List]

# Experiment 4 - All features using MinMax-Normalization for numeric and OneHot for categorical. No custom features.
exp4_train = X_train[minMax_feature_list +  OneHot_Encoding_Feature_List]
exp4_test = X_test[minMax_feature_list + OneHot_Encoding_Feature_List]

# Experiment 5 - The kitchen sink: Include everything to try to get the best performance possible.
exp5_train = X_train[numeric_features + standardScaler_Feature_List +  minMax_feature_list + OneHot_Encoding_Feature_List + targetEncoded_feature_List]
exp5_test = X_test[numeric_features + standardScaler_Feature_List +  minMax_feature_list + OneHot_Encoding_Feature_List + targetEncoded_feature_List]

# Experiment 6 - Only custom features. How good can you get with your own custom feature set?
exp6_train = X_train[custom_features]
exp6_test = X_test[custom_features]

# Experiment 7 - Only categorical features using one of the encodings.
exp7_train = X_train[OneHot_Encoding_Feature_List]
exp7_test = X_test[OneHot_Encoding_Feature_List]

# Experiment 8 - Only numerical features using one of the encodings.
exp8_train = X_train[standardScaler_Feature_List]
exp8_test = X_test[standardScaler_Feature_List]


experiments_train = [exp1_train, exp2_train, exp3_train, exp4_train, exp5_train, exp6_train, exp7_train, exp8_train]
experiments_test = [exp1_test, exp2_test, exp3_test, exp4_test, exp5_test, exp6_test, exp7_test, exp8_test]

In [16]:
# Global values so that it is easier to save data <-- otherwise function parameters get too long
experiment_name, columns, model_type = [], [], []
feature_list, epochs, hyperparameters_list = [], [], []
training_accuracy, testing_accuracy = [], []
training_pr_auc, testing_pr_auc = [], []
training_precision, testing_precision = [], []
training_recall, testing_recall = [], []

In [17]:
model = LogisticRegression(max_iter=1000).fit(experiments_train[0], y_train)
print(model.predict(experiments_train[0])[0:10]) # Clearly it predicts 0 or 1 so there is no need for a threshold
model = DecisionTreeClassifier().fit(experiments_train[0], y_train)
print(model.predict(experiments_train[0])[0:10]) # Clearly it predicts 0 or 1 so there is no need for a threshold
model = MLPClassifier().fit(experiments_train[0], y_train)
print(model.predict(experiments_train[0])[0:10]) # Clearly it predicts 0 or 1 so there is no need for a threshold
model = GradientBoostingClassifier().fit(experiments_train[0], y_train)
print(model.predict(experiments_train[0])[0:10]) # Clearly it predicts 0 or 1 so there is no need for a threshold


[0 0 1 0 1 1 1 1 1 0]
[0 0 1 0 1 1 1 1 1 0]
[0 0 1 0 1 1 1 1 1 0]
[0 0 1 0 1 1 1 1 1 0]


In [18]:
# Storing model results
def log_results(name, hyperparameters, model, X_train, y_train, X_test, y_test):
    # Saving the experiment name 
    experiment_name.append(name)
    # Saving the columns used
    feature_list.append(";".join(X_train.columns))
    # Saving the Hyperparameters
    hyperparameters_list.append(hyperparameters)
    # Saving Accuracies
    training_accuracy.append(accuracy_score(y_train, model.predict(X_train)))
    testing_accuracy.append(accuracy_score(y_test, model.predict(X_test)))
    # Getting Results
    tr1, tr2, tr3 = precision_recall_curve(y_train, model.predict(X_train)) # training_precisions, training_recalls, training_thresholds
    te1, te2, te3 = precision_recall_curve(y_test, model.predict(X_test)) # testing_precisions, testing_recalls, testing_thresholds
    # Saving PR-AUC
    training_pr_auc.append(auc(tr2,  tr1)) 
    testing_pr_auc.append(auc(te2, te1))
    # Saving Precision
    training_precision.append(precision_score(y_train, model.predict(X_train), zero_division=0))
    testing_precision.append(precision_score(y_test, model.predict(X_test), zero_division=0))
    # Saving Recall
    training_recall.append(recall_score(y_train, model.predict(X_train)))
    testing_recall.append(recall_score(y_test, model.predict(X_test)))

In [20]:
# Function for the LogisticRegression model
def LogisticReg(name, X_train, y_train, X_test, y_test):
    model_type.append("LogisticRegression")
    model = LogisticRegression(solver="lbfgs", max_iter=600).fit(X_train, y_train) # 1
    epochs.append(model.n_iter_)
    log_results(name, "solver=lbfgs max_iter=600", model, X_train, y_train, X_test, y_test)

    model_type.append("LogisticRegression")
    model = LogisticRegression(solver="newton-cg", max_iter=600).fit(X_train, y_train) # 2
    epochs.append(model.n_iter_)
    log_results(name, "solver=newton-cg max_iter=600", model, X_train, y_train, X_test, y_test)

    model_type.append("LogisticRegression")
    model = LogisticRegression(solver="liblinear", max_iter=600).fit(X_train, y_train) # 3
    epochs.append(model.n_iter_)
    log_results(name, "solver=liblinear max_iter=600", model, X_train, y_train, X_test, y_test)

    model_type.append("LogisticRegression")
    model = LogisticRegression(solver="sag", max_iter=5000).fit(X_train, y_train) # 4
    epochs.append(model.n_iter_)
    log_results(name, "solver=sag max_iter=600", model, X_train, y_train, X_test, y_test)

    model_type.append("LogisticRegression")
    model = LogisticRegression(solver="saga", max_iter=5000).fit(X_train, y_train) # 5
    epochs.append(model.n_iter_)
    log_results(name, "solver=sag max_iter=600", model, X_train, y_train, X_test, y_test)
    
    model_type.append("LogisticRegression")
    model = LogisticRegression(solver="newton-cg", penalty="l2", max_iter=600).fit(X_train, y_train) # 6
    epochs.append(model.n_iter_)
    log_results(name, "solver=newton-cg, penalty=l2 max_iter=600", model, X_train, y_train, X_test, y_test)
    
    model_type.append("LogisticRegression")
    model = LogisticRegression(solver="newton-cg", penalty="none", max_iter=600).fit(X_train, y_train) # 7
    epochs.append(model.n_iter_)
    log_results(name, "solver=newton-cg, penalty=none max_iter=600", model, X_train, y_train, X_test, y_test)
    
    model_type.append("LogisticRegression")
    model = LogisticRegression(solver="lbfgs", penalty="l2", max_iter=600).fit(X_train, y_train) # 8
    epochs.append(model.n_iter_)
    log_results(name, "solver=lbfgs, penalty=l2, max_iter=600", model, X_train, y_train, X_test, y_test)

    model_type.append("LogisticRegression")
    model = LogisticRegression(solver="lbfgs", penalty="none", max_iter=600).fit(X_train, y_train) # 9
    epochs.append(model.n_iter_)
    log_results(name, "solver=lbfgs, penalty=none, max_iter=600", model, X_train, y_train, X_test, y_test)

    model_type.append("LogisticRegression")
    model = LogisticRegression(solver="lbfgs", penalty="l2", tol=1e-10, max_iter=600).fit(X_train, y_train) # 10
    epochs.append(model.n_iter_)
    log_results(name, "solver=lbfgs, penalty=l2, tol=1e-10, max_iter=600", model, X_train, y_train, X_test, y_test)

In [21]:
def DecisionTree(name, X_train, y_train, X_test, y_test):
    model_type.append("DecisionTreeClassifier")
    epochs.append("N/A")
    model = DecisionTreeClassifier(criterion="gini").fit(X_train, y_train) # 1
    log_results(name, "criterion=gini", model, X_train, y_train, X_test, y_test)

    model_type.append("DecisionTreeClassifier")
    epochs.append("N/A")
    model = DecisionTreeClassifier(criterion="entropy").fit(X_train, y_train) # 2
    log_results(name, "criterion=entropy", model, X_train, y_train, X_test, y_test)

    model_type.append("DecisionTreeClassifier")
    epochs.append("N/A")
    model = DecisionTreeClassifier(criterion="gini", splitter="best").fit(X_train, y_train) # 3
    log_results(name, "criterion=gini, splitter=best", model, X_train, y_train, X_test, y_test)

    model_type.append("DecisionTreeClassifier")
    epochs.append("N/A")
    model = DecisionTreeClassifier(criterion="gini", splitter="random").fit(X_train, y_train) # 4
    log_results(name, "criterion=gini, splitter=random", model, X_train, y_train, X_test, y_test)

    model_type.append("DecisionTreeClassifier")
    epochs.append("N/A")
    model = DecisionTreeClassifier(criterion="gini", splitter="random", max_features="auto").fit(X_train, y_train) # 5
    log_results(name, "criterion=gini, splitter=random, max_features=auto", model, X_train, y_train, X_test, y_test)

    model_type.append("DecisionTreeClassifier")
    epochs.append("N/A")
    model = DecisionTreeClassifier(criterion="gini", splitter="random", max_features="sqrt").fit(X_train, y_train) # 6
    log_results(name, "criterion=gini, splitter=random, max_features=sqrt", model, X_train, y_train, X_test, y_test)


    model_type.append("DecisionTreeClassifier")
    epochs.append("N/A")
    model = DecisionTreeClassifier(criterion="gini", splitter="random", max_features="log2").fit(X_train, y_train) # 7
    log_results(name, "criterion=gini, splitter=random, max_features=log2", model, X_train, y_train, X_test, y_test)

    model_type.append("DecisionTreeClassifier")
    epochs.append("N/A")
    model = DecisionTreeClassifier(criterion="gini", splitter="random", max_features="sqrt", max_depth=300).fit(X_train, y_train) # 8
    log_results(name, "criterion=gini, splitter=random, max_features=sqrt, max_depth=300", model, X_train, y_train, X_test, y_test)

    model_type.append("DecisionTreeClassifier")
    epochs.append("N/A")
    model = DecisionTreeClassifier(criterion="gini", splitter="random", max_features="sqrt", max_depth=300, min_samples_split=2).fit(X_train, y_train) # 9
    log_results(name, "criterion=gini, splitter=random, max_features=sqrt, max_depth=300, min_samples_split=2", model, X_train, y_train, X_test, y_test)

    model_type.append("DecisionTreeClassifier")
    epochs.append("N/A")
    model = DecisionTreeClassifier(criterion="gini", splitter="random", max_features="sqrt", max_depth=300, min_samples_split=2, min_samples_leaf=1).fit(X_train, y_train) # 10
    log_results(name, "criterion=gini, splitter=random, max_features=sqrt, max_depth=300, min_samples_split=2, min_samples_leaf=1", model, X_train, y_train, X_test, y_test)


In [22]:
def MLP(name, X_train, y_train, X_test, y_test):
    model_type.append("MLPClassifier")
    epochs.append("N/A")
    model = MLPClassifier(solver="adam", max_iter=3000).fit(X_train, y_train) # 1
    log_results(name, "solver=adam, max_iter=3000", model, X_train, y_train, X_test, y_test)

    model_type.append("MLPClassifier")
    epochs.append("N/A")
    model = MLPClassifier(solver="adam", activation="relu", max_iter=3000).fit(X_train, y_train) # 2
    log_results(name, "solver=adam, activation=relu, max_iter=3000", model, X_train, y_train, X_test, y_test)

    model_type.append("MLPClassifier")
    epochs.append("N/A")
    model = MLPClassifier(solver="adam", activation="relu", hidden_layer_sizes=(3, 2), max_iter=3000).fit(X_train, y_train) # 3
    log_results(name, "solver=adam, activation=relu, hidden_layer_sizes=(3, 2), max_iter=3000", model, X_train, y_train, X_test, y_test)

    model_type.append("MLPClassifier")
    epochs.append("N/A")
    model = MLPClassifier(solver="adam", activation="relu", hidden_layer_sizes=(4, 1),  max_iter=3000).fit(X_train, y_train) # 4
    log_results(name, "solver=adam, activation=relu, hidden_layer_sizes=(4, 1), max_iter=3000", model, X_train, y_train, X_test, y_test)

    model_type.append("MLPClassifier")
    epochs.append("N/A")
    model = MLPClassifier(solver="adam", activation="relu", hidden_layer_sizes=(100, 100), max_iter=3000).fit(X_train, y_train) # 5
    log_results(name, "solver=adam, activation=relu, hidden_layer_sizes=(100, 100), max_iter=3000", model, X_train, y_train, X_test, y_test)

    model_type.append("MLPClassifier")
    epochs.append("N/A")
    model = MLPClassifier(solver="adam", activation="relu", hidden_layer_sizes=(100, 100), alpha=0.001, max_iter=3000).fit(X_train, y_train) # 6
    log_results(name, "solver=adam, activation=relu, hidden_layer_sizes=(100, 100), alpha=0.001, max_iter=3000", model, X_train, y_train, X_test, y_test)

    model_type.append("MLPClassifier")
    epochs.append("N/A")
    model = MLPClassifier(solver="adam", activation="relu", hidden_layer_sizes=(100, 100), alpha=0.001, learning_rate="constant", max_iter=3000).fit(X_train, y_train) # 7
    log_results(name, "solver=adam, activation=relu, hidden_layer_sizes=(100, 100), alpha=0.001, learning_rate=constant, max_iter=3000", model, X_train, y_train, X_test, y_test)

    model_type.append("MLPClassifier")
    epochs.append("N/A")
    model = MLPClassifier(solver="adam", activation="relu", hidden_layer_sizes=(100, 100), alpha=0.001, learning_rate="adaptive", max_iter=3000).fit(X_train, y_train) # 8
    log_results(name, "solver=adam, activation=relu, hidden_layer_sizes=(100, 100), alpha=0.001, learning_rate=adaptive, max_iter=3000", model, X_train, y_train, X_test, y_test)
    
    model_type.append("MLPClassifier")
    epochs.append("N/A")
    model = MLPClassifier(solver="adam", activation="relu", hidden_layer_sizes=(100, 100), alpha=0.001, learning_rate="invscaling", max_iter=3000).fit(X_train, y_train) # 9
    log_results(name, "solver=adam, activation=relu, hidden_layer_sizes=(100, 100), alpha=0.001, learning_rate=invscaling, max_iter=3000", model, X_train, y_train, X_test, y_test)
    
    model_type.append("MLPClassifier")
    epochs.append("N/A")
    model = MLPClassifier(solver="adam", activation="relu", hidden_layer_sizes=(100, 100), alpha=0.001, learning_rate="constant", shuffle=False, max_iter=3000).fit(X_train, y_train) # 10
    log_results(name, "solver=adam, activation=relu, hidden_layer_sizes=(100, 100), alpha=0.001, learning_rate=constant, shuffle=False, max_iter=3000", model, X_train, y_train, X_test, y_test)

In [23]:
def GradientBoost(name, X_train, y_train, X_test, y_test):
    model_type.append("GradientBoostClassifier")
    epochs.append("N/A")
    model = GradientBoostingClassifier(loss="deviance").fit(X_train, y_train) # 1
    log_results(name, "loss=deviance", model, X_train, y_train, X_test, y_test)

    model_type.append("GradientBoostClassifier")
    epochs.append("N/A")
    model = GradientBoostingClassifier(loss="exponential").fit(X_train, y_train) # 2
    log_results(name, "loss=exponential", model, X_train, y_train, X_test, y_test)

    model_type.append("GradientBoostClassifier")
    epochs.append("N/A")
    model = GradientBoostingClassifier(loss="deviance", criterion="friedman_mse").fit(X_train, y_train) # 3
    log_results(name, "loss=deviance, criterion=friedman_mse", model, X_train, y_train, X_test, y_test)

    model_type.append("GradientBoostClassifier")
    epochs.append("N/A")
    model = GradientBoostingClassifier(loss="deviance", criterion="squared_error").fit(X_train, y_train) # 4
    log_results(name, "loss=deviance, criterion=squared_error", model, X_train, y_train, X_test, y_test)

    model_type.append("GradientBoostClassifier")
    epochs.append("N/A")
    model = GradientBoostingClassifier(loss="deviance", criterion="friedman_mse", min_samples_split=2).fit(X_train, y_train) # 5
    log_results(name, "loss=deviance, criterion=friedman_mse, min_samples_split=2", model, X_train, y_train, X_test, y_test)

    model_type.append("GradientBoostClassifier")
    epochs.append("N/A")
    model = GradientBoostingClassifier(loss="deviance", criterion="friedman_mse", min_samples_split=9).fit(X_train, y_train) # 6
    log_results(name, "loss=deviance, criterion=friedman_mse, min_samples_split=9", model, X_train, y_train, X_test, y_test)

    model_type.append("GradientBoostClassifier")
    epochs.append("N/A")
    model = GradientBoostingClassifier(loss="deviance", criterion="friedman_mse", min_samples_split=9, min_samples_leaf=1).fit(X_train, y_train) # 7
    log_results(name, "loss=deviance, criterion=friedman_mse, min_samples_split=9, min_samples_leaf=1", model, X_train, y_train, X_test, y_test)

    model_type.append("GradientBoostClassifier")
    epochs.append("N/A")
    model = GradientBoostingClassifier(loss="deviance", criterion="friedman_mse", min_samples_split=9, min_samples_leaf=3).fit(X_train, y_train) # 8
    log_results(name, "loss=deviance, criterion=friedman_mse, min_samples_split=9, min_samples_leaf=3", model, X_train, y_train, X_test, y_test)

    model_type.append("GradientBoostClassifier")
    epochs.append("N/A")
    model = GradientBoostingClassifier(loss="deviance", criterion="friedman_mse", min_samples_split=9, min_samples_leaf=3, learning_rate=0.3).fit(X_train, y_train) # 9
    log_results(name, "loss=deviance, criterion=friedman_mse, min_samples_split=9, min_samples_leaf=3, learning_rate=0.28", model, X_train, y_train, X_test, y_test)

    model_type.append("GradientBoostClassifier")
    epochs.append("N/A")
    model = GradientBoostingClassifier(loss="deviance", criterion="friedman_mse", min_samples_split=9, min_samples_leaf=3, learning_rate=0.28).fit(X_train, y_train) # 10
    log_results(name, "loss=deviance, criterion=friedman_mse, min_samples_split=9, min_samples_leaf=3, learning_rate=0.28", model, X_train, y_train, X_test, y_test)

In [24]:
# Models
def experiment(name, X_train, y_train, X_test, y_test):
    # model_type.append("LogisticRegression")
    # model = LogisticRegression(solver="lbfgs", penalty="l2", max_iter=600).fit(X_train, y_train)
    # epochs.append(model.n_iter_)
    # log_results(name, "solver=lbfgs, penalty=l2, max_iter=600", model, X_train, y_train, X_test, y_test)
    LogisticReg(name, X_train, y_train, X_test, y_test)
    
    # model_type.append("DecisionTreeClassifier")
    # epochs.append("N/A")
    # model = DecisionTreeClassifier(criterion="gini", splitter="random", max_features="sqrt", max_depth=300, min_samples_split=2, min_samples_leaf=1).fit(X_train, y_train)
    # log_results(name, "criterion=gini, splitter=random, max_features=sqrt, max_depth=300, min_samples_split=2, min_samples_leaf=1", model, X_train, y_train, X_test, y_test)
    DecisionTree(name, X_train, y_train, X_test, y_test)

    # model_type.append("MLPClassifier")
    # epochs.append("N/A")
    # model = MLPClassifier(solver="adam", activation="relu", hidden_layer_sizes=(100, 100), alpha=0.001, learning_rate="constant", shuffle=False, max_iter=2000).fit(X_train, y_train)
    # log_results(name, "solver=adam, activation=relu, hidden_layer_sizes=(100, 100), alpha=0.001, learning_rate=constant, shuffle=False, max_iter=1000", model, X_train, y_train, X_test, y_test)
    MLP(name, X_train, y_train, X_test, y_test)

    # model_type.append("GradientBoostClassifier")
    # epochs.append("N/A")
    # model = GradientBoostingClassifier(loss="deviance", criterion="friedman_mse", min_samples_split=9, min_samples_leaf=3, learning_rate=0.28).fit(X_train, y_train)
    # log_results(name, "loss=deviance, criterion=friedman_mse, min_samples_split=9, min_samples_leaf=3, learning_rate=0.28", model, X_train, y_train, X_test, y_test)
    GradientBoost(name, X_train, y_train, X_test, y_test)

In [26]:
# Question 1: Run Experiments <--> This runs all 8 experiments with 4 models and logs all the results into the HW_3.csv
exp_names = ["Numerical-OneHot", "StandardScaler-OneHot", "StandardScaler-TargetEncoding", "MinMax-OneHot", "KitchenSink", "Custom-Features", "Categorical", "Numerical"]
for i, j, k, w in zip(range(1, 9), exp_names, experiments_train, experiments_test):
        experiment(f"Experiment-{i}-"+j, k, y_train, w, y_test)

config = {
        "Experiment Name": experiment_name,
        "Feature_List": feature_list,
        "Model": model_type,
        "Epochs": epochs,
        "Hyperparameters": hyperparameters_list,
        "Training Accuracy": training_accuracy,
        "Testing Accuracy": testing_accuracy,
        "Training PR-AUC": training_pr_auc,
        "Testing PR-AUC": testing_pr_auc,
        "Training Precision":training_precision,
        "Testing Precision": testing_precision,
        "Training Recall": training_recall,
        "Testing Recall": testing_recall
}
answers = pd.DataFrame(config)
answers.to_csv("HW_3.csv")

# Wandb for extra credit
wandb.init(project="hw03")
wandb.config = {
        "Experiment Name": experiment_name,
        "Feature_List": feature_list,
        "Model": model_type,
        "Epochs": epochs,
        "Hyperparameters": hyperparameters_list,
        "Training Accuracy": training_accuracy,
        "Testing Accuracy": testing_accuracy,
        "Training PR-AUC": training_pr_auc,
        "Testing PR-AUC": testing_pr_auc,
        "Training Precision":training_precision,
        "Testing Precision": testing_precision,
        "Training Recall": training_recall,
        "Testing Recall": testing_recall
}
wandb.log(wandb.config)
#If you run this cell becareful because it takes awhile to run all the experiments 

In [228]:
# Question 2.1: What is the best set of features and parameters for Logistic Regression in terms of Test PR-AUC. Is this the same as the best in terms of accuracy? 
# Discuss why you think this experiment showed the best results
# 2.1 - Looked for results in csv to make sure
best_test_pr_auc, row, max_testacc, max_testacc_row = 0, 0, 0, 0
for i in range(len(answers["Model"])):
    if answers["Model"][i] == "LogisticRegression":
        if answers["Testing PR-AUC"][i] > best_test_pr_auc:
            best_test_pr_auc = answers["Testing PR-AUC"][i]
            row = i
        if answers["Testing Accuracy"][i] > max_testacc:
            max_testacc = answers["Testing Accuracy"][i]
            max_testacc_row = i
            
print("Best Test PR-AUC for Logistic Regression is in", experiment_name[row], " with a Test PR-AUC of", testing_pr_auc[row], "in row", row, "of the csv")
print("Best Test Accuracy for Logistic Regression is", max_testacc, "on row", max_testacc_row, "in the csv")
print("Best hyperameters for test accuracy with Logistic Regression is", answers["Hyperparameters"][max_testacc_row])

Best Test PR-AUC for Logistic Regression is in Experiment-3-StandardScaler-TargetEncoding  with a Test PR-AUC of 0.928914881677986 in row 80 of the csv
Best Test Accuracy for Logistic Regression is 0.8804347826086957 on row 80 in the csv
Best hyperameters for test accuracy with Logistic Regression is solver=lbfgs max_iter=600


In [229]:
# Question 2.2: If features are well normalized, we can get a sense of feature importance by looking at the absolute value of the weights of each feature.
# Print out the highest 5 weights by absolute value. Which features are these? Discuss what this set of 5 features has on the model.  
train = X_train[standardScaler_Feature_List+minMax_feature_list+OneHot_Encoding_Feature_List+targetEncoded_feature_List]
test = X_test[standardScaler_Feature_List+minMax_feature_list+OneHot_Encoding_Feature_List+targetEncoded_feature_List]
model = LogisticRegression().fit(train, y_train)

topfive = [0, 0, 0, 0, 0]
for i in range(0, 5):
    high = -1
    for j in range(len(model.coef_[0])):
        if high < abs(model.coef_[0][j]):
            high = abs(model.coef_[0][j])
            if j not in topfive:
                topfive[i] = j
print(topfive) 
print(model.feature_names_in_[topfive])

[25, 24, 15, 12, 2]
['ST_Slope-OneHot-Flat' 'ST_Slope-OneHot-Up' 'ChestPainType-OneHot-ASY'
 'Sex-OneHot-M' 'Cholesterol-StandardScaler']


In [230]:
# Question 2.3: Train a model with just those top 5 features. What is the Test Accuracy and PR-AUC?
print(model.predict(test)[0:5]) # Threshold not needed because all values are 0, 1
print("Testing Accuracy:", accuracy_score(y_test, model.predict(test)))
te1, te2, te3 = precision_recall_curve(y_test, model.predict(test)) # testing_precisions, testing_recalls, testing_thresholds 
print("Test PR-AUC:", auc(te2, te1))


[0 0 1 1 0]
Testing Accuracy: 0.8586956521739131
Test PR-AUC: 0.9192069552343288


### Question 2: Analyze Logistic Regression Experiments (4 points)

- Question 2.1: What is the best set of features and parameters for Logistic Regression in terms of Test PR-AUC. Is this the same as the best in terms of accuracy? Discuss why you think this experiment showed the best results.
- Question 2.2: If features are well normalized, we can get a sense of feature importance by looking at the absolute value of the weights of each feature. Print out the highest 5 weights by absolute value. Which features are these? Discuss what this set of 5 features has on the model.
- Question 2.3: Train a model with just those top 5 features. What is the Test Accuracy and PR-AUC?
- Question 2.4: Describe the custom features you added. Why did you pick these?


- Answer 2.1: The best set of features for Logistic Regression is in experiment 3 which is all the numerical features in StandardScaler and all categorical features in TargetEncoding. The best parameters for Logistic Regression is solver=lbfgs and max_iter=600. The best Test PR-AUC for Logisitic Regression is the same one as the one with best Test accuracy. This is probably because PR-AUC is calculated with y_test and model.predict(X_test), which basically means the higher the accuracy the less false positives or false negatives leading to a higher PR-AUC score.  
Answer proof in cells above.

- Answer 2.2: The highest 5 weights are ['ST_Slope_Flat' 'ST_Slope_Up' 'ChestPainType_ASY' 'Sex_M'
 'Cholesterol-Scaled']. These 5 features affect the model's prediction if the person has heart disease or not the most because they 
 were given the highest weight in the equation for predicting heart disease by the model.
Answer proof in cells above.

- Answer 2.3: Testing Accuracy: 0.8586956521739131 <----> Test PR-AUC: 0.9192069552343288 Answer proof in cells above.

- Answer 2.4: I made custom features for Age, MaxHR, Oldpeak using binary, bucketizing and cross of two custom features.
I picked the custom features because after using df.corr(), I saw that these features have a high correlation with heart disease and I thought making custom features from these features would improve the accuracy.

In [231]:
# Qyestion 3 - Proof of answers
row, pr_auc = [0, 0, 0], [0, 0, 0]
for i in range(len(answers["Model"])):
    if answers["Model"][i] == "DecisionTreeClassifier":
        if answers["Testing PR-AUC"][i] > pr_auc[0]:
            pr_auc[0] = answers["Testing PR-AUC"][i]
            row[0] = i
    if answers["Model"][i] == "MLPClassifier":
        if answers["Testing PR-AUC"][i] > pr_auc[1]:
            pr_auc[1] = answers["Testing PR-AUC"][i]
            row[1] = i
    if answers["Model"][i] == "GradientBoostClassifier":
        if answers["Testing PR-AUC"][i] > pr_auc[2]:
            pr_auc[2] = answers["Testing PR-AUC"][i]
            row[2] = i

print(answers["Experiment Name"][row[0]])
print(answers["Hyperparameters"][row[0]])
print("Test PR-AUC for DecisionTreeClassifier", pr_auc[0])
print()
print(answers["Experiment Name"][row[1]])
print(answers["Hyperparameters"][row[1]])
print("Test PR-AUC for Neural Net classifier", pr_auc[1])
print()
print(answers["Experiment Name"][row[2]])
print(answers["Hyperparameters"][row[2]])
print("Test PR-AUC for Gradient Boost", pr_auc[2])

Experiment-5-KitchenSink
criterion=gini, splitter=random, max_features=sqrt, max_depth=300, min_samples_split=2
Test PR-AUC for DecisionTreeClassifier 0.9118249798903715

Experiment-4-MinMax-OneHot
solver=adam, activation=relu, max_iter=3000
Test PR-AUC for Neural Net classifier 0.9426702390627115

Experiment-2-StandardScaler-OneHot
loss=deviance, criterion=friedman_mse, min_samples_split=9, min_samples_leaf=3, learning_rate=0.28
Test PR-AUC for Gradient Boost 0.9423497029856234


### Question 3: Analyze Other Experiments (4 points)

- Question 3.1: What is the best set of features and parameters for Decision Tree classifier in terms of Test PR-AUC.  Discuss why you think this experiment showed the best results.
- Question 3.2: What is the best set of features and parameters for Neural Net classifier in terms of Test PR-AUC.  Discuss why you think this experiment showed the best results.
- Question 3.3: What is the best set of features and parameters for another chosen model in terms of Test PR-AUC.  Discuss why you think this experiment showed the best results.

- Answer 3.1: The best set of features for best Test PR-AUC for the DecisionTreeClassifier model was in experiment 5 which uses all the features, including StandardScale, MinMax, OneHotEncoding, TargetEncoding and all the custom features. The best set of parameters for best Test PR-AUC using the DecisionTreeClassifiers model was criterion=gini, splitter=random, max_features=sqrt, max_depth=300, min_samples_split=2. Experiment 5 
showed the best result for the DecisionTreeClassifier model because this experiment has all the features in it. Due to the fact
that experiment 5 contains all the features it most likely played a large part in increasing the accuracy of the decisions made by each tree
in the DecisionTreeClassifier model.

- Answer 3.2: Suprisingly for the neural net classifier in terms of Test PR-AUC, experiment 4 had the highest. The features found the best were numerical columns as MinMax and categorical columns as oneHot encoding. The parameters found the best were solver=adam, activation=relu, max_iter=3000. I was suprised that the Neural net classifier had the highest Test PR-AUC on experiment 4 but if I had to point a reason as to why this occured it would be probably be because the numeric columns are represented as MinMax. In fact, it is more suspicious that the MinMax
features affected the model since experiment 4 is the only experiment where the only representation of numerical features was MinMax and the fact that the only other feature type was OneHotEncoding which also used in other experiments and did not produce similar results.

- Answer 3.3: The GradientBoostClassifier model in terms of Test PR-AUC was best in experiment 2. In experiment 2, the features that
where used was the StandardScaled features and the OneHotEncoded features.
The parameters utilized in GradientBoostClassifier was loss=deviance, criterin=friedman_mse, min_samples_split=9, min_samples_leaf=3, learning_rate=0.28. I think the GradientBoostClassifier model had the best Test PR-AUC in this case because of the StandardScaler feature,
in the csv after looking at the scores it seemed like the GradientBoostClassifer did significantly better where the StandardScaled features
was utilized.

In [282]:
# Question 4.1
print("Model:", answers["Model"][24], "\tHyperparameters:", answers["Hyperparameters"][24])
print("Testing Recall:", answers["Testing Recall"][24]) 
print("Testing Precision", answers["Testing Precision"][24]) 
print("Training Precision", answers["Training Precision"][24], "\n")
# Question 4.2
print("Model:", answers["Model"][26], "\tHyperparameters:", answers["Hyperparameters"][26])
print("Testing Precision", answers["Testing Precision"][26])
print("Testing Recall:", answers["Testing Recall"][26]) 
print("Training Recall", answers["Training Recall"][24])
# I got the row by looking at the csv

Model: MLPClassifier 	Hyperparameters: solver=adam, activation=relu, hidden_layer_sizes=(100, 100), max_iter=3000
Testing Recall: 0.9345794392523364
Testing Precision 0.8403361344537815
Training Precision 0.8071278825995807 

Model: MLPClassifier 	Hyperparameters: solver=adam, activation=relu, hidden_layer_sizes=(100, 100), alpha=0.001, learning_rate=constant, max_iter=3000
Testing Precision 0.9518072289156626
Testing Recall: 0.7383177570093458
Training Recall 0.9600997506234414


### Question 4: Achieving particular peformance charactaristics (2 points)
- Question 4.1: From your trained models, can you produce a classifier that has around 90% recall on the test set? How? What is the precision of this model?
- Question 4.2: From your trained models, can you produce a classifier that has around 90% precision on the test set? How? What is the recall of this model?

- Answer 4.1: Yes I can produce a classifier with around 90% recall on the test set. I obtained this recall by utilizing the MLPClassifier with the parameters solver=adam, activation=relu, hidden_layer_sizes=(100, 100), max_iter=3000. The testing precision of this model was 0.8403361344537815 and the training precision of this model was 0.8071278825995807 .

- Answer 4.2: Yes I can get a classifier with around 90% precision on the test set. I obtained this precision by utiziling the MLPClassfier with
the parameters solver=adam, activation=relu, hidden_layer_sizes=(100, 100), alpha=0.001, learning_rate=constant, max_iter=3000. The testing recall of this model was 0.0.7383177570093458 and the training recall was 0.9600997506234414.

In [288]:
# Qyestion 5 - Proof of answer
sum = [[0,0], [0,0], [0,0], [0,0]]
for i in range(len(answers["Model"])):
    if answers["Model"][i] == "LogisticRegression":
        sum[0][0] += answers["Training Accuracy"][i]
        sum[0][1] += answers["Testing Accuracy"][i]
    elif answers["Model"][i] == "DecisionTreeClassifier":
        sum[1][0] += answers["Training Accuracy"][i]
        sum[1][1] += answers["Testing Accuracy"][i]
    elif answers["Model"][i] == "MLPClassifier":
        sum[2][0] += answers["Training Accuracy"][i]
        sum[2][1] += answers["Testing Accuracy"][i]
    elif answers["Model"][i] == "GradientBoostClassifier":
        sum[3][0] += answers["Training Accuracy"][i]
        sum[3][1] += answers["Testing Accuracy"][i]


model_list = ["LogisticRegression", "DecisionTreeClassifier", "MLPClassifier", "GradientBoostClassifier"]
print("Training Accuracy, Testing Accuracy")
for i, j in zip(sum, model_list):
    print(f"{j}: {i}")

print("\nSum of all Accuracy")
for i, j in zip(sum, model_list):
    print(f"{j}: {i[0]+i[1]}")

Training Accuracy, Testing Accuracy
LogisticRegression: [67.33514986376022, 65.47826086956528]
DecisionTreeClassifier: [76.29427792915529, 60.994565217391354]
MLPClassifier: [71.13215258855584, 63.74456521739137]
GradientBoostClassifier: [72.91825613079021, 66.679347826087]

Sum of all Accuracy
LogisticRegression: 132.8134107333255
DecisionTreeClassifier: 137.28884314654664
MLPClassifier: 134.8767178059472
GradientBoostClassifier: 139.5976039568772


### Question 5: Taking a step back (2 points)
- Question 5.1: What is your best performing model over all experiments? Describe why you think this was the best? What is the PR-AUC and Accuracy of this model? Feel free to share your results on discord, the best model in the class will receive extra credit of 3 points!
- Question 5.2: Discuss your thoughts on this process? What surprised you? What was hard or tedious? What did you learn?

- Answer 5.1: The best performing model was either the DecisionTreeClassider model or the GradientBoostClassifier. The DecisionTreeClassifier
model had the highest training accuracy, while the GradientBoostClassifier had the highest combined training and test accuracy.
- Answer 5.2: The process was tedious but I think it could have been a lot easier if the homework's questions instructions was guided in a different order. I also learned a lot about saving results to a csv and wandb because I've never really saved my data to a file from python in any other class.

In [6]:
d = pd.read_csv("HW_3.csv")
columns  = [i for i in d.head(0)]
columns.remove("Unnamed: 0")
columns.remove("Epochs")
mx, index = 0, 0
for i in range(len(d["Model"])):
    if d["Model"][i] == "LogisticRegression":
        if mx < d["Training Accuracy"][i] + d["Testing Accuracy"][i]:
            mx = d["Training Accuracy"][i] + d["Testing Accuracy"][i]
            index = i
    elif d["Model"][i] == "DecisionTreeClassifier":
        if mx < d["Training Accuracy"][i] + d["Testing Accuracy"][i]:
            mx = d["Training Accuracy"][i] + d["Testing Accuracy"][i]
            index = i
    elif d["Model"][i] == "MLPClassifier":
        if mx < d["Training Accuracy"][i] + d["Testing Accuracy"][i]:
            mx = d["Training Accuracy"][i] + d["Testing Accuracy"][i]
            index = i
    elif d["Model"][i] == "GradientBoostClassifier":
        if mx < d["Training Accuracy"][i] + d["Testing Accuracy"][i]:
            mx = d["Training Accuracy"][i] + d["Testing Accuracy"][i]
            index = i

for i in columns:
    print(f"{i}:", d[i][index])

Experiment Name: Experiment-2-StandardScaler-OneHot
Feature_List: Sex-OneHot-M;Sex-OneHot-F;ChestPainType-OneHot-NAP;ChestPainType-OneHot-ASY;ChestPainType-OneHot-TA;ChestPainType-OneHot-ATA;RestingECG-OneHot-Normal;RestingECG-OneHot-LVH;RestingECG-OneHot-ST;ExerciseAngina-OneHot-N;ExerciseAngina-OneHot-Y;ST_Slope-OneHot-Down;ST_Slope-OneHot-Up;ST_Slope-OneHot-Flat;Age-StandardScaler;RestingBP-StandardScaler;Cholesterol-StandardScaler;FastingBS-StandardScaler;MaxHR-StandardScaler;Oldpeak-StandardScaler
Model: GradientBoostClassifier
Hyperparameters: loss=deviance, criterion=friedman_mse, min_samples_split=9, min_samples_leaf=3, learning_rate=0.28
Training Accuracy: 0.998637602
Testing Accuracy: 0.902173913
Training PR-AUC: 0.998756219
Testing PR-AUC: 0.942349703
Training Precision: 0.997512438
Testing Precision: 0.923809524
Training Recall: 1.0
Testing Recall: 0.906542056
